File: sdks/ruby/conformance_runner.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Ruby SDK conformance runner.
#
# Reads every *.json case in the shared conformance suite
# (../../conformance/suite/*.json by default) and runs each one through
# the pure-Ruby implementation. Prints PASS/FAIL per case and a summary.
# Exits 0 on full pass, 1 on any failure.

require "json"

$LOAD_PATH.unshift(File.expand_path("lib", __dir__))
require "crowdcontrol"

DEFAULT_SUITE = File.expand_path("../../conformance/suite", __dir__)

def load_case(path)
  JSON.parse(File.read(path))
end

def run_case(c)
  default_effect = c["default_effect"] || "allow"
  unless %w[allow deny].include?(default_effect)
    return [false, "unknown default_effect #{default_effect.inspect}"]
  end

  begin
    eng = CrowdControl.from_source([c["policy"]], default_effect: default_effect)
  rescue CrowdControl::ParseError => e
    return [false, "parse error: #{e.message}"]
  end

  results  = eng.evaluate(c["input"] || {})
  expected = (c["expect"] || {})["decisions"] || []

  if results.length != expected.length
    summary = results.map { |r| "[#{r.rule}/#{r.kind} passed=#{r.passed}]" }.join(" ")
    return [false, "expected #{expected.length} decisions, got #{results.length} (results: #{summary})"]
  end

  expected.each_with_index do |want, i|
    got = results[i]
    return [false, "decision[#{i}]: rule = #{got.rule.inspect}, want #{want['rule'].inspect}"] if got.rule != want["rule"]
    return [false, "decision[#{i}] (#{got.rule}): kind = #{got.kind.inspect}, want #{want['kind'].inspect}"] if got.kind != want["kind"]
    return [false, "decision[#{i}] (#{got.rule}): passed = #{got.passed}, want #{want['passed']}"] if got.passed != want["passed"]

    if want.key?("message_exact") && want["message_exact"] != "" && got.message != want["message_exact"]
      return [false, "decision[#{i}] (#{got.rule}): message = #{got.message.inspect}, want exact #{want['message_exact'].inspect}"]
    end
    if want.key?("message_contains") && want["message_contains"] != "" && !got.message.include?(want["message_contains"])
      return [false, "decision[#{i}] (#{got.rule}): message = #{got.message.inspect}, want contains #{want['message_contains'].inspect}"]
    end
  end

  [true, ""]
end

def main(argv)
  suite_dir = argv[0] || DEFAULT_SUITE
  verbose   = argv.include?("-v") || argv.include?("--verbose")
  filter_i  = argv.index("-f") || argv.index("--filter")
  filter    = filter_i ? argv[filter_i + 1] : ""

  unless File.directory?(suite_dir)
    warn "suite dir not found: #{suite_dir}"
    return 2
  end

  files = Dir.children(suite_dir)
              .select { |f| f.end_with?(".json") }
              .reject { |f| File.directory?(File.join(suite_dir, f)) }
              .sort
              .map { |f| File.join(suite_dir, f) }

  if files.empty?
    warn "no conformance cases in #{suite_dir}"
    return 2
  end

  passed = 0
  failed = 0

  files.each do |path|
    begin
      c = load_case(path)
    rescue StandardError => e
      puts "FAIL: #{File.basename(path)} — load error: #{e.message}"
      failed += 1
      next
    end

    name = c["name"] || File.basename(path, ".json")
    next if filter && filter != "" && !name.include?(filter)

    ok, msg = run_case(c)
    if ok
      passed += 1
      puts "PASS: #{name}" if verbose
    else
      failed += 1
      puts "FAIL: #{name} — #{msg}"
    end
  end

  puts
  puts "#{passed} passed, #{failed} failed"
  failed > 0 ? 1 : 0
end

exit(main(ARGV)) if __FILE__ == $PROGRAM_NAME

File: sdks/python/conformance_runner.py
#!/usr/bin/env python3
"""Python SDK conformance runner.

Reads every case file in the shared conformance suite
(``../../conformance/suite/*.json``), runs it through the Python
implementation, and verifies the results match the expected decisions.

Usage::

    python conformance_runner.py [SUITE_DIR]

Exits 0 on full pass, 1 on any failure.
"""

from __future__ import annotations

import argparse
import json
import os
import sys
from typing import Any

# Allow running directly from the package dir without install.
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import crowdcontrol  # noqa: E402

DEFAULT_SUITE = os.path.normpath(
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "conformance", "suite")
)


def load_case(path: str) -> dict[str, Any]:
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def run_case(case: dict[str, Any]) -> tuple[bool, str]:
    default_effect = case.get("default_effect") or "allow"
    if default_effect not in ("allow", "deny"):
        return False, f"unknown default_effect {default_effect!r}"

    try:
        eng = crowdcontrol.from_source([case["policy"]], default_effect=default_effect)
    except crowdcontrol.ParseError as e:
        return False, f"parse error: {e}"

    results = eng.evaluate(case.get("input", {}))
    expected = case.get("expect", {}).get("decisions", [])

    if len(results) != len(expected):
        summary = " ".join(
            f"[{r.rule}/{r.kind} passed={r.passed}]" for r in results
        )
        return False, f"expected {len(expected)} decisions, got {len(results)} (results: {summary})"

    for i, want in enumerate(expected):
        got = results[i]
        if got.rule != want["rule"]:
            return False, f"decision[{i}]: rule = {got.rule!r}, want {want['rule']!r}"
        if got.kind != want["kind"]:
            return False, f"decision[{i}] ({got.rule}): kind = {got.kind!r}, want {want['kind']!r}"
        if got.passed != want["passed"]:
            return False, f"decision[{i}] ({got.rule}): passed = {got.passed}, want {want['passed']}"
        if "message_exact" in want and want["message_exact"] != "" and got.message != want["message_exact"]:
            return False, (
                f"decision[{i}] ({got.rule}): message = {got.message!r}, "
                f"want exact {want['message_exact']!r}"
            )
        if "message_contains" in want and want["message_contains"] != "" and want["message_contains"] not in got.message:
            return False, (
                f"decision[{i}] ({got.rule}): message = {got.message!r}, "
                f"want contains {want['message_contains']!r}"
            )

    return True, ""


def main() -> int:
    parser = argparse.ArgumentParser(description="Run the CrowdControl conformance suite against the Python SDK")
    parser.add_argument("suite", nargs="?", default=DEFAULT_SUITE, help="path to conformance/suite directory")
    parser.add_argument("-v", "--verbose", action="store_true", help="print passing cases too")
    parser.add_argument("-f", "--filter", default="", help="only run cases whose name contains this substring")
    args = parser.parse_args()

    if not os.path.isdir(args.suite):
        print(f"suite dir not found: {args.suite}", file=sys.stderr)
        return 2

    files = sorted(
        os.path.join(args.suite, f)
        for f in os.listdir(args.suite)
        if f.endswith(".json") and not os.path.isdir(os.path.join(args.suite, f))
    )
    if not files:
        print(f"no conformance cases in {args.suite}", file=sys.stderr)
        return 2

    passed = 0
    failed = 0
    for path in files:
        try:
            case = load_case(path)
        except Exception as e:
            print(f"FAIL: {os.path.basename(path)} — load error: {e}")
            failed += 1
            continue

        name = case.get("name") or os.path.splitext(os.path.basename(path))[0]
        if args.filter and args.filter not in name:
            continue

        ok, msg = run_case(case)
        if ok:
            passed += 1
            if args.verbose:
                print(f"PASS: {name}")
        else:
            failed += 1
            print(f"FAIL: {name} — {msg}")

    print()
    print(f"{passed} passed, {failed} failed")
    return 1 if failed > 0 else 0


if __name__ == "__main__":
    sys.exit(main())
